lint/config.go

```diff
@@ -109,55 +89,42 @@ func ReadConfigsYAML(f io.Reader) (Configs, error) {
 	return c, nil
 }
 
-// GetRuleConfig returns a RuleConfig that matches the given path and rule.
-// Returns an error if a config is not found for the path.
-func (c Configs) GetRuleConfig(path string, rule RuleName) (result RuleConfig, err error) {
-	err = fmt.Errorf("failed to find a config for path %q", path)
-	for _, cfg := range c {
-		if cfg.match(path) {
-			err = nil
-			if r, ok := cfg.getRuleConfig(rule); ok {
-				result = result.withOverride(r)
+// IsRuleEnabled returns true if a rule is enabled by the configs.
+func (configs Configs) IsRuleEnabled(rule string, path string) bool {
+	enabled := true
+	for _, c := range configs {
+		if c.matchPath(path) {
+			if matchRule(rule, c.DisabledRules...) {
+				enabled = false
+			}
+			if matchRule(rule, c.EnabledRules...) {
+				enabled = true
 			}
-		}
-	}
-	return
-}
-
-func (c Config) getRuleConfig(rule RuleName) (RuleConfig, bool) {
-	for r := rule; ; r = r.parent() {
-		if ruleConfig, ok := c.RuleConfigs[string(r)]; ok {
-			return ruleConfig, true
-		}
-
-		if r == "" {
-			break
 		}
 	}
 
-	return RuleConfig{}, false
+	return enabled
 }
 
-// withOverride returns a copy of r, overridden with non-zero values in r2
-func (r RuleConfig) withOverride(r2 RuleConfig) RuleConfig {
-	r.Disabled = r2.Disabled
-
-	if r2.Category != "" {
-		r.Category = r2.Category
+func (c Config) matchPath(path string) bool {
+	if matchPath(path, c.ExcludedPaths...) {
+		return false
 	}
-
-	return r
+	return len(c.IncludedPaths) == 0 || matchPath(path, c.IncludedPaths...)
 }
 
-// match returns if a Config matches path based on its included and excluded paths
-func (c Config) match(path string) bool {
-	for _, pattern := range c.ExcludedPaths {
-		if matched, err := doublestar.Match(pattern, path); matched || err != nil {
-			return false
+func matchPath(path string, pathPatterns ...string) bool {
+	for _, pattern := range pathPatterns {
+		if matched, _ := doublestar.Match(pattern, path); matched {
+			return true
 		}
 	}
-	for _, pattern := range c.IncludedPaths {
-		if matched, err := doublestar.Match(pattern, path); matched && err == nil {
+	return false
+}
+
+func matchRule(rule string, rulePrefixes ...string) bool {
+	for _, prefix := range rulePrefixes {
+		if prefix == "all" || prefix == rule || strings.HasPrefix(rule, prefix+"::") {
```

Comment:
It need not block this PR, but can we have the logic be "matches any segment" rather than "matches prefix"?

For example, I would like someone to be able to write `(-- api-linter: http-uri=disabled --)` at the top of a file and disable `http-uri` checks from _any_ AIP.

Reply:
I added segment matching in the configs, but not in the file comments, which we will address it in another PR. Created a issue #326. Thanks!